feat(table): enable multiple data rows

andrewseguin opened the pull request:
Introduces a new table input `multiTemplateDataRows` which allows the table to have multiple data rows per data object.

**Old paradigm**

Each data object passed to the table is rendered as a data row. Since the mapping is 1:1, the data array itself can be used with the `IterableDiffer` to determine which rows need to change.

**New paradigm**

Each data object passed to the table may be rendered by one or more rows. The `IterableDiffer` cannot use the data array itself anymore since it will not correctly understand add/remove/move operations in the rows. 

Instead, the `IterableDiffer` will be diffed with a `DataRow` object that represents a pair containing a data object and row template. When the data changes, the list of `DataRow` objects will be constructed and diffed to get the list of changes.

**Caveats for discussion**

1. The row context previously contained an `index` value that represented both the row index and data object index (they were the same. Now, that `index` value represents the index of the data. To get the row index, there is an additional context property called `rowIndex`.

_Should it be swapped such that `index` is the row index and the data index can be stored in `dataIndex`?_ 

2. Similarly, the `trackBy` function receives an index value. Previously the index matched both the data and row. Now, its explicitly the data index.

_Should `trackBy` receive the row index?_

jelbourn commented:
For the index stuff: I think that, when in multiple-row mode, you should have separate `renderIndex` and `dataIndex` properties for both the context and `trackBy`; I wouldn't make either just `index` because it will always be ambiguous. However, when in the standard mode, it should stay just `index` that that will by far be the most common case.

jelbourn commented on src/cdk/table/table.ts:
```diff
@@ -262,8 +306,9 @@ export class CdkTable<T> implements CollectionViewer, OnInit, AfterContentChecke
 
-    // TODO(andrewseguin): Setup a listener for scrolling, emit the calculated view to viewChange
-    this._dataDiffer = this._differs.find([]).create(this._trackByFn);
+    this._dataDiffer = this._differs.find([]).create((_i: number, dataRow: DataRow<T>) => {
```
Add comment explaining what is happening here? It's not quite clear to me

jelbourn commented on src/cdk/table/table.ts:
```diff
@@ -393,6 +439,56 @@ export class CdkTable<T> implements CollectionViewer, OnInit, AfterContentChecke
+   * row definitions. If the previous list already contained a particular pair, it should be reused
+   * so that the differ equates their references.
+   */
+  private _getDataRows(): DataRow<T>[] {
```
Having both `_getDataRows` and `_getDataRowsForData` is confusing

andrewseguin replied:
Agreed, struggled coming up with a better name for these. One returns the list of all the `DataRow` objects to render, while one returns the list of `DataRow` objects for a particular data object.

Any ideas?

jelbourn replied:
I think `_getAllRenderRows` and `_getRenderRowsForData` would be good

andrewseguin commented on src/cdk/table/table.ts:
```diff
@@ -98,6 +99,16 @@ export const CDK_TABLE_TEMPLATE = `
 
+/**
+ * Set of properties that represent a single rendered row. Acts as the row's identity for the
+ * `IterableDiffer`.
```
Added more to the comment, not sure whether its too verbose.

andrewseguin commented on src/cdk/table/table.ts:
```diff
@@ -162,6 +177,19 @@ export class CdkTable<T> implements CollectionViewer, OnInit, AfterContentChecke
+   * `CdkRowDef<T>` object. With the two keys, the cache points to a `DataRow<T>` object that
+   * contains the pair.
+   */
+  private _cachedDataRowsMap = new Map<T, Map<CdkRowDef<T>, DataRow<T>>>();
```
Unfortunately not.  `T` may be a primitive value which would be an invalid `WeakMap` key.

andrewseguin commented on src/cdk/table/table.ts:
```diff
@@ -162,6 +177,19 @@ export class CdkTable<T> implements CollectionViewer, OnInit, AfterContentChecke
+   * `CdkRowDef<T>` object. With the two keys, the cache points to a `DataRow<T>` object that
+   * contains the pair.
+   */
+  private _cachedDataRowsMap = new Map<T, Map<CdkRowDef<T>, DataRow<T>>>();
```
Each `T` may map into multiple `DataRow<T>` objects since each object represents a pair of data and row template.

This setup allows us to uniquely identify these pairs so that they can be re-used if they show up again between renders. This is necessary for the `IterableDiffer` to understand the correct operations, e.g. when a pairing is moved.

andrewseguin commented on src/cdk/table/table.ts:
```diff
@@ -209,6 +237,22 @@ export class CdkTable<T> implements CollectionViewer, OnInit, AfterContentChecke
+   * defined in the table, or otherwise the default row which does not have a when predicate.
+   */
+  @Input()
+  get enableRowMultiplex(): boolean { return this._enableRowMultiplex; }
```
Perhaps it should be `multiTemplateDataRows` since I intend to allow multiple header and footer rows which would not depend on this input.

andrewseguin commented:
Making the index and naming changes; will add `pr: needs review` label when I pushed the latest

andrewseguin commented:
Ready for review

andrewseguin commented on src/cdk/table/table.spec.ts:
```diff
@@ -941,30 +979,55 @@ class BooleanRowCdkTableApp {
   dataSource: FakeDataSource = new FakeDataSource();
   columnsToRender = ['column_a', 'column_b', 'column_c'];
+  isIndex1Columns = ['index1Column'];
+  hasC3Columns = ['c3Column'];
```
Renamed to `columnsForIsIndex1Row` and `columnsForHasC3Row`. Not totally sure thats better :)

andrewseguin commented on src/cdk/table/table.ts:
```diff
@@ -93,11 +95,37 @@ export const CDK_TABLE_TEMPLATE = `
+ * @docs-private
+ */
+export interface RowContext<T>
+    extends CdkCellOutletMultiRowContext<T>, CdkCellOutletRowContext<T> { }
```
Tried that but typescript didn't like it. Perhaps I'm doing something incorrectly?

Here's a typescript playground that shows what I tried: https://www.typescriptlang.org/play/#src=interface%20A%20%7B%0D%0A%20%20%20%20a1%3A%20number%3B%0D%0A%20%20%20%20a2%3A%20number%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20B%20%7B%0D%0A%20%20%20%20b1%3A%20number%3B%0D%0A%20%20%20%20b2%3A%20number%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20ExtendedInterface%20extends%20A%2C%20B%20%7B%20%7D%3B%0D%0Alet%20value%3A%20ExtendedInterface%3B%0D%0Avalue.a1%20%3D%201%3B%0D%0Avalue.b1%20%3D%201%3B%0D%0A%0D%0Atype%20UnionInterface%20%3D%20A%20%7C%20B%3B%0D%0Alet%20unionValue%3A%20UnionInterface%3B%0D%0AunionValue.a1%20%3D%201%3B%0D%0AunionValue.b1%20%3D%201%3B

andrewseguin commented on src/cdk/table/table.ts:
```diff
@@ -93,11 +95,37 @@ export const CDK_TABLE_TEMPLATE = `
+ */
+export interface RenderRow<T> {
+  data: T;
+  index: number;
```
Renamed to `dataIndex`

andrewseguin commented on src/cdk/table/table.ts:
```diff
@@ -394,6 +463,56 @@ export class CdkTable<T> implements CollectionViewer, OnInit, AfterContentChecke
+
+      this._cachedRenderRowsMap.set(data, new WeakMap());
+      for (let j = 0; j < dataRowsForData.length; j++) {
+        let dataRow = dataRowsForData[j];
```
Ack thanks, missed these

andrewseguin commented:
Thanks, should have done a better job of renaming dataRow to renderRow. Forgot to find/replace that lowercase one version

andrewseguin commented:
Note to caretaker, will require BUILD change to add `":cdk_coercion"` to the `cdk_table` BUILD rule. 

See change 194450307 for patch.

andrewseguin commented on src/cdk/table/table.ts:
```diff
@@ -394,6 +462,56 @@ export class CdkTable<T> implements CollectionViewer, OnInit, AfterContentChecke
+
+    // For each data object, get the list of rows that should be rendered, represented by the
+    // respective `RenderRow` object which is the pair of `data` and `CdkRowDef`.
+    for (let i = 0; i < this._data.length; i++) {
```
Need to check/handle if empty

andrewseguin commented:
Removing `pr: needs review` label. Found that having homogeneous data was not correctly being handled between diffs due to caching a `RenderRow` for each data object rather than per rendered row.

Writing up some tests and will add the label and a comment when ready again.

andrewseguin commented:
Ready again for review. Added docs on the table's implementation process for rendering rows. Renamed the input to `multiTemplateDataRows` since the input will only affect data rows

jelbourn commented on src/cdk/table/table.ts:
```diff
@@ -394,6 +464,66 @@ export class CdkTable<T> implements CollectionViewer, OnInit, AfterContentChecke
+  }
+
+  /**
+   * Returns a list of `RenderRow<T>` for the provided data object and any `CdkRowDef` objects that
```
Prefer "Gets" to "Returns` in descriptions

jelbourn commented on src/cdk/table/table.ts:
```diff
@@ -532,27 +649,38 @@ export class CdkTable<T> implements CollectionViewer, OnInit, AfterContentChecke
    */
-  _getRowDef(data: T, i: number): CdkRowDef<T> {
-    if (this._rowDefs.length == 1) { return this._rowDefs[0]; }
+  _getRowDefs(data: T, i: number): CdkRowDef<T>[] {
```
Still make this more specific than `i`?

jelbourn commented on src/cdk/table/table.spec.ts:
```diff
@@ -523,24 +563,97 @@ describe('CdkTable', () => {
+        ]);
+      });
+
+      it('should have the correct data and row indicies when data is homogeneous', () => {
```
"Homogeneous" doesn't really communicate the same thing as "Data contains multiple occurrences of the same object instance"

andrewseguin commented:
Thanks - made the changes
